android/source/src/main/kotlin/com/gu/source/components/buttons/SourceIconButton.kt

```diff
@@ -0,0 +1,277 @@
+package com.gu.source.components.buttons
+
+import androidx.annotation.VisibleForTesting
+import androidx.compose.foundation.background
+import androidx.compose.foundation.border
+import androidx.compose.foundation.layout.*
+import androidx.compose.foundation.shape.CircleShape
+import androidx.compose.material.icons.Icons
+import androidx.compose.material3.Icon
+import androidx.compose.material3.IconButton
+import androidx.compose.material3.IconButtonDefaults
+import androidx.compose.runtime.Composable
+import androidx.compose.ui.Alignment
+import androidx.compose.ui.Modifier
+import androidx.compose.ui.graphics.ImageBitmap
+import androidx.compose.ui.graphics.painter.Painter
+import androidx.compose.ui.graphics.vector.ImageVector
+import androidx.compose.ui.unit.dp
+import com.gu.source.Source
+import com.gu.source.daynight.AppColourMode
+import com.gu.source.icons.Check
+import com.gu.source.presets.palette.Neutral38
+import com.gu.source.theme.LocalSourceTheme
+import com.gu.source.theme.ReaderRevenueTheme
+import com.gu.source.theme.SourceCoreTheme
+import com.gu.source.utils.PhoneBothModePreviews
+
+/**
+ * Source icon-only button. This variant allows providing the icon as a composable slot.
+ *
+ * @param size The size of the button. Reflects the prominence of the action.
+ * @param priority The priority of the button. Informs users of how important an action is.
+ * @param onClick The action to perform when the button is clicked.
+ * @param modifier The modifier to apply to the button.
+ * @param theme Optional [Source.Theme] to apply to the button. If not provided, the current theme
+ * from [LocalSourceTheme] will be used.
+ *
+ * Unless using the button as a standalone component, it is recommended to wrap the whole
+ * screen/sheet with [SourceCoreTheme] or [ReaderRevenueTheme] to provide consistent theme to all
+ * child components once.
+ * @param icon The icon to display in the button. Use a material [Icon] component to display the
+ * icon.
+ */
+@Composable
+fun SourceBaseIconButton(
+    size: SourceButton.Size,
+    priority: SourceButton.Priority,
+    onClick: () -> Unit,
+    modifier: Modifier = Modifier,
+    theme: Source.Theme? = null,
+    icon: @Composable (Modifier) -> Unit = {},
+) {
+    val appliedTheme = theme ?: LocalSourceTheme.current
+
+    require(
```

Comment:
Didn't know about this! Neat way of forcing devs to provide you the correct parameters. It would be nice if kotlin had a simple way to enforce this using the type checker at compile time. If you had one parameter of sort of ButtonThemeAndPriority which was a sealed class with a value for theme and priority you could enforce this by not having a subclass for ReaderRevenueSecondary. But I can imagine this becoming a huge faff in the number of priorities or themes increases.

Something like. Maybe worth considering.

```kotlin
sealed class ButtonThemeAndPriority(
  priority: SourceButton.Priority,
  theme: Source.Theme,
) {

data object ReaderRevenuePrimary : ButtonThemeAndPriority(
  priority = SourceButton.Priority.Primary,
  theme = Source.Theme.ReaderRevenue,
)

}
```

Reply:
I considered merging theme and priority. The main reason is that it's preferred that theme is not explicitly required on all components. That's why it supports reading theme from composition local. 
Using the composition local allows, e.g. if we're building a metering screen, to wrap the whole screen in `ReaderRevenueScope` so all buttons and any other components are automatically correctly themed.
But using the composition local also means that we can't rely on a combined theme + priority parameter.

Reply:
Makes sense to me! Cheers for the explanation!